shell: add plugin to detect first task exit

garlick opened the pull request:
I thought this was going to be a throw-away demo of a plugin that detects the first task exit and logs an exec eventlog entry, but after adding a timeout that throws a fatal job exception 30s after that, it seems like it might actually be sufficient to check off that feature for now.

So posting this as a WIP to get design feedback.

Possibly the throw-away name should be changed, as `doom` was chosen becuase it had morbid connotations and was short :-)

Is it OK for the shell to log the event directly to the eventlog or should it be notifying the `job-exec` service?

I sure am liking the ability we have to implement stuff like this as plugins.

garlick commented:
Force pushed a bit of cleanup here.

grondo commented:
I tested this up to 256 shells on fluke with no detected fallout, so this seems good to go in!

I was initially wondering if the eventlog event should be named `first-task-exit` vs `task-exit`, but I think since it is the first `task-exit` event in the eventlog, the "first" is implied!

grondo commented:
Slurm also has an option to kill a job immediately on the first nonzero task exit. It seems like this plugin could be extended to support that feature if necessary.

garlick commented:
> Seems fine as the internal name for the plugin, but probably the shell option to set the timeout should be renamed to something that better indicates what it does.

>  Slurm also has an option to kill a job immediately on the first nonzero task exit.

OK I'll throw out some probably horrible ideas and defer to others like you and @ryanday36 for better suggestions?
```
$ flux mini run -ojobicide.exit-timeout=10s  ...
$ flux mini run -ojobicide.exit-on-error ...
```

grondo commented:
I like just `-o exit-timeout=30s`.

Then add documentation to `JOBSPEC OPTIONS` section of `flux-shell(1)` and call it good for now?
This fixes a real problem so it would be cool to get it in our next tag.

garlick commented:
Just hitting the road so I'd you'd like to force push something here feel
free. O/w I'll make changes tonight or tomorrow.  Thanks!

On Fri, May 28, 2021, 8:46 AM Mark Grondona ***@***.***>
wrote:

> I like just -o exit-timeout=30s.
>
> Then add documentation to JOBSPEC OPTIONS section of flux-shell(1) and
> call it good for now?
> This fixes a real problem so it would be cool to get it in our next tag.
>
> —
> You are receiving this because you authored the thread.
> Reply to this email directly, view it on GitHub
> <https://github.com/flux-framework/flux-core/pull/3681#issuecomment-850509768>,
> or unsubscribe
> <https://github.com/notifications/unsubscribe-auth/AABJPW6QHF6CBJU5JEQ5P3DTP63HBANCNFSM45SQVZDA>
> .
>

grondo commented on src/shell/doom.c:
```diff
@@ -0,0 +1,243 @@
+    if (!doom->done) {
+        json_t *entry;
+
+        if (!(entry = doom_eventlog_entry_create (task)))
```
Hm, maybe it would be better to create the eventlog entry on the rank 0 shell? This avoids strange timestamps when clocks have drifted:

```ƒ(s=392,d=0,builddir) grondo@fluke2:~/git/f.git$ flux mini run -N 392 -n 392 -vvv sh -c 'sleep .$((392 - $FLUX_TASK_RANK))'
jobid: ƒ3s3MjkBZ
0.000s: job.submit {"userid":6885,"urgency":16,"flags":0}
0.014s: job.depend
0.014s: job.priority {"priority":16}
0.021s: job.alloc {"annotations":{"sched":{"resource_summary":"rank[0-391]/core0"}}}
0.382s: job.start
0.022s: exec.init
0.024s: exec.starting
0.327s: exec.shell.init {"leader-rank":0,"size":392,"service":"6885-shell-6320950345728"}
0.379s: exec.shell.start {"task-count":392}
-10.770s: exec.shell.task-exit {"localid":0,"rank":391,"state":"Exited","pid":1327786,"wait_status":0,"signaled":0,"exitcode":0}
1.335s: exec.complete {"status":0}
1.335s: exec.done
1.335s: job.finish {"status":0}
1.339s: job.release {"ranks":"all","final":true}
1.342s: job.free
1.342s: job.clean
```

garlick commented:
OK, force pushed with:
* Eventlog "envelope" is created on rank 0 for uniform timestamps
* Option is `exit-timeout`
* Added shell(1) blurb
* Rebased on current master

grondo commented on src/shell/doom.c:
```diff
@@ -0,0 +1,231 @@
+    const char *fsd = NULL;
+    double timeout;
+
+    if (flux_shell_getopt_unpack (shell, "exit-timeout", "s", &fsd) < 0)
```
Should an integer or real work here as well? (implied seconds), maybe unpack as a json_t object and use json_string_value()?

garlick replied:
Good idea - that would avoid the somewhat generic error you get now with an integer argument.

garlick replied:
Oh, just realized it already does since I was using the FSD parser

https://flux-framework.readthedocs.io/projects/flux-rfc/en/latest/spec_23.html

I'll push a fix to the value error message though.

garlick replied:
Oh wait, no it doesn't get past jansson does it.  Fixing.

garlick replied:
OK, force pushed a change and some more tests.

garlick commented:
Oops, I was using `json_real_value()` where I should have been using `json_number_value()` and had inverted the logic in the test for it.  Pushed again.

grondo commented:
That json_t to FSD snippet might eventually be useful in libutil as this is a common use case. Too bad we can't extend `unpack()` with a special symbol to unpack a number or string as FSD into a double...

grondo reviewed (approved):
LGTM, nice work!

grondo commented:
Ok to set MWP?

garlick commented:
Thanks, I'll go ahead and set it.
